File: src/ObjectManager/Object.Tes/FilePacks/Records/345-CLOT.Clothing.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class CLOTRecord : Record, IHaveEDID, IHaveMODL
    {
        // TESX
        public struct DATAField
        {
            public enum CLOTType
            {
                Pants = 0, Shoes, Shirt, Belt, Robe, R_Glove, L_Glove, Skirt, Ring, Amulet
            }

            public int Value;
            public float Weight;
            //
            public int Type;
            public short EnchantPts;

            public DATAField(UnityBinaryReader r, int dataSize, GameFormatId format)
            {
                if (format == GameFormatId.TES3)
                {
                    Type = r.ReadLEInt32();
                    Weight = r.ReadLESingle();
                    Value = r.ReadLEInt16();
                    EnchantPts = r.ReadLEInt16();
                    return;
                }
                Value = r.ReadLEInt32();
                Weight = r.ReadLESingle();
                Type = 0;
                EnchantPts = 0;
            }
        }

        public class INDXFieldGroup
        {
            public override string ToString() => $"{INDX.Value}: {BNAM.Value}";
            public INTVField INDX;
            public STRVField BNAM;
            public STRVField CNAM;
        }

        public override string ToString() => $"CLOT: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public MODLGroup MODL { get; set; } // Model Name
        public STRVField FULL; // Item Name
        public DATAField DATA; // Clothing Data
        public FILEField ICON; // Male Icon
        public STRVField ENAM; // Enchantment Name
        public FMIDField<SCPTRecord> SCRI; // Script Name
        // TES3
        public List<INDXFieldGroup> INDXs = new List<INDXFieldGroup>(); // Body Part Index (Moved to Race)
        // TES4
        public UI32Field BMDT; // Clothing Flags
        public MODLGroup MOD2; // Male world model (optional)
        public MODLGroup MOD3; // Female biped (optional)
        public MODLGroup MOD4; // Female world model (optional)
        public FILEField? ICO2; // Female icon (optional)
        public IN16Field? ANAM; // Enchantment points (optional)

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID":
                case "NAME": EDID = new STRVField(r, dataSize); return true;
                case "MODL": MODL = new MODLGroup(r, dataSize); return true;
                case "MODB": MODL.MODBField(r, dataSize); return true;
                case "MODT": MODL.MODTField(r, dataSize); return true;
                case "FULL":
                case "FNAM": FULL = new STRVField(r, dataSize); return true;
                case "DATA":
                case "CTDT": DATA = new DATAField(r, dataSize, format); return true;
                case "ICON":
                case "ITEX": ICON = new FILEField(r, dataSize); return true;
                case "INDX": INDXs.Add(new INDXFieldGroup { INDX = new INTVField(r, dataSize) }); return true;
                case "BNAM": INDXs.Last().BNAM = new STRVField(r, dataSize); return true;
                case "CNAM": INDXs.Last().CNAM = new STRVField(r, dataSize); return true;
                case "ENAM": ENAM = new STRVField(r, dataSize); return true;
                case "SCRI": SCRI = new FMIDField<SCPTRecord>(r, dataSize); return true;
                case "BMDT": BMDT = new UI32Field(r, dataSize); return true;
                case "MOD2": MOD2 = new MODLGroup(r, dataSize); return true;
                case "MO2B": MOD2.MODBField(r, dataSize); return true;
                case "MO2T": MOD2.MODTField(r, dataSize); return true;
                case "MOD3": MOD3 = new MODLGroup(r, dataSize); return true;
                case "MO3B": MOD3.MODBField(r, dataSize); return true;
                case "MO3T": MOD3.MODTField(r, dataSize); return true;
                case "MOD4": MOD4 = new MODLGroup(r, dataSize); return true;
                case "MO4B": MOD4.MODBField(r, dataSize); return true;
                case "MO4T": MOD4.MODTField(r, dataSize); return true;
                case "ICO2": ICO2 = new FILEField(r, dataSize); return true;
                case "ANAM": ANAM = new IN16Field(r, dataSize); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/345-CLOT.Clothing.swift
//
//  CLOTRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class CLOTRecord: Record, IHaveEDID, IHaveMODL {
    // TESX
    public struct DATAField {
        public enum CLOTType: UInt32 {
            case pants = 0, shoes, shirt, belt, robe, r_glove, l_glove, skirt, ring, amulet
        }

        public let value: Int32
        public let weight: Float
        //
        public let type: Int32
        public let enchantPts: Int16

        init(_ r: BinaryReader, _ dataSize: Int, _ format: GameFormatId) {
            guard format != .TES3 else {
                type = r.readLEInt32()
                weight = r.readLESingle()
                value = Int32(r.readLEInt16())
                enchantPts = r.readLEInt16()
                return
            }
            value = r.readLEInt32()
            weight = r.readLESingle()
            type = 0
            enchantPts = 0
        }
    }

    public class INDXFieldGroup: CustomStringConvertible {
        public var description: String { return "\(INDX): \(BNAM ?? "")" }
        public var INDX: INTVField
        public var BNAM: STRVField!
        public var CNAM: STRVField!
        
        init(INDX: INTVField) {
            self.INDX = INDX
        }
    }

    public override var description: String { return "CLOT: \(EDID)" }
    public var EDID: STRVField = STRVField_empty  // Editor ID
    public var MODL: MODLGroup? = nil // Model Name
    public var FULL: STRVField! // Item Name
    public var DATA: DATAField! // Clothing Data
    public var ICON: FILEField! // Male Icon
    public var ENAM: STRVField! // Enchantment Name
    public var SCRI: FMIDField<SCPTRecord>! // Script Name
    // TES3
    public var INDXs = [INDXFieldGroup]() // Body Part Index (Moved to Race)
    // TES4
    public var BMDT: UI32Field!  // Clothing Flags
    public var MOD2: MODLGroup!  // Male world model (optional)
    public var MOD3: MODLGroup!  // Female biped (optional)
    public var MOD4: MODLGroup!  // Female world model (optional)
    public var ICO2: FILEField? = nil // Female icon (optional)
    public var ANAM: IN16Field? = nil // Enchantment points (optional)

    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID",
             "NAME": EDID = r.readSTRV(dataSize)
        case "MODL": MODL = MODLGroup(r, dataSize)
        case "MODB": MODL!.MODBField(r, dataSize)
        case "MODT": MODL!.MODTField(r, dataSize)
        case "FULL",
             "FNAM": FULL = r.readSTRV(dataSize)
        case "DATA",
             "CTDT": DATA = DATAField(r, dataSize, format)
        case "ICON",
             "ITEX": ICON = r.readSTRV(dataSize)
        case "INDX": INDXs.append(INDXFieldGroup(INDX: r.readINTV(dataSize)))
        case "BNAM": INDXs.last!.BNAM = r.readSTRV(dataSize)
        case "CNAM": INDXs.last!.CNAM = r.readSTRV(dataSize)
        case "ENAM": ENAM = r.readSTRV(dataSize)
        case "SCRI": SCRI = FMIDField<SCPTRecord>(r, dataSize)
        case "BMDT": BMDT = r.readT(dataSize)
        case "MOD2": MOD2 = MODLGroup(r, dataSize)
        case "MO2B": MOD2.MODBField(r, dataSize)
        case "MO2T": MOD2.MODTField(r, dataSize)
        case "MOD3": MOD3 = MODLGroup(r, dataSize)
        case "MO3B": MOD3.MODBField(r, dataSize)
        case "MO3T": MOD3.MODTField(r, dataSize)
        case "MOD4": MOD4 = MODLGroup(r, dataSize)
        case "MO4B": MOD4.MODBField(r, dataSize)
        case "MO4T": MOD4.MODTField(r, dataSize)
        case "ICO2": ICO2 = r.readSTRV(dataSize)
        case "ANAM": ANAM = r.readT(dataSize)
        default: return false
        }
        return true
    }
}
